src/Microsoft.Extensions.Diagnostics.HealthChecks.Abstractions/HealthCheckResult.cs

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.Diagnostics.HealthChecks
+{
+    /// <summary>
+    /// Represents the result of a health check.
+    /// </summary>
+    public struct HealthCheckResult
+    {
+        private static readonly IReadOnlyDictionary<string, object> _emptyReadOnlyDictionary = new Dictionary<string, object>();
+
+        private string _description;
+        private IReadOnlyDictionary<string, object> _data;
+
+        /// <summary>
+        /// Gets a <see cref="HealthCheckStatus"/> value indicating the status of the component that was checked.
+        /// </summary>
+        public HealthCheckStatus Status { get; }
+
+        /// <summary>
+        /// Gets an <see cref="Exception"/> representing the exception that was thrown when checking for status (if any).
+        /// </summary>
+        /// <remarks>
+        /// This value is expected to be 'null' if <see cref="Status"/> is <see cref="HealthCheckStatus.Healthy"/>.
+        /// </remarks>
+        public Exception Exception { get; }
+
+        /// <summary>
+        /// Gets a human-readable description of the status of the component that was checked.
+        /// </summary>
+        public string Description => _description ?? string.Empty;
+
+        /// <summary>
+        /// Gets additional key-value pairs describing the health of the component.
+        /// </summary>
+        public IReadOnlyDictionary<string, object> Data => _data ?? _emptyReadOnlyDictionary;
```

Comment:
The big reason to avoid string,string is that you may frequently want to emit numeric data (amount of memory consumed, etc.), and you want that preserved as a number in the JSON payload

Reply:
Yes! When a problem does occur. A structured error like a structured log is far more useful in automated systems to diagnose the issue

Reply:
> Or we can keep it <string, object> and just be ready to say no to those requests when they come.

Yeah, that's my approach. You want to serialize it, you can build your own middleware ;)